### ISSUE TITLE
Protein Viewer

### ISSUE THREAD

**Issue by KianBadie** [caltechlibrary/cell-atlas#12]
Issue to discuss integration of a protein viewer into the Cell Atlas.

**Comment by KianBadie** [comment]
@coiko I decided to create an issue for this topic now to decrease the back and forth of email. You mentioned that the protein viewer will be accessed through the "Structure" field in the video citation sections for modals. When the protein viewer is opened, did you imagine it taking up the same amount of space that the videos/sliders do, and it being in the same position? Or did you imagine it kind of being like the enlarged sliders in modals, where a new window of its own takes over the modal? Or was it something else?

**Comment by coiko** [comment]
@KianBadie I like the idea of being similar to the enlarged slider in the modals.  Thanks!

**Comment by KianBadie** [comment]
@coiko The demo pages are up on our testing site! The Mol* and PV demos are available at these URLs:
- https://caltechlibrary.github.io/cell-atlas/molstar-demo
- https://caltechlibrary.github.io/cell-atlas/pv-demo

Like we talked about in our meeting, the viewers are opened by clicking the "Open In Viewer" text next to the pdb in the modal. There is currently no way to reshow the regular modal content once the viewer is opened, but I left it like that because I felt like feeling out the viewer functionality was the main purpose. If you resize the window, there might be some buggy characteristics as well.

As you probably already know, clicking the wrench in the Mol* viewer shows all the viewer settings (like coloring and model type). For the PV viewer, one model type and one color type may be selected together at a time. I included all the built in coloring schemes that I could currently find in PV. Unfortunately, there is no official list of properties to color by. However, I feel like I included the obvious ones (if not all the ones PV has available), and there are a few more properties I will try to check to see if PV will accept them as coloring schemes. 

Please let me know if you have any questions or if anything prevents you from comparing the viewers properly!

**Comment by coiko** [comment]
@KianBadie This is great - thanks!  I'll play around with them in the next few days and let you know what I think.

**Comment by KianBadie** [comment]
@coiko To follow up on my message from last week, I am going to go ahead and assume that all the coloring options for PV are covered now. I looked a bit into the implementation of the PV package to see if maybe there were some properties to color by that weren't documented/obvious, however it is starting to feel unlikely that there is anything more than what I included. So I think it should be assumed for now that there are no further PV options and that what is in the demo is what will be available as the default.

**Comment by coiko** [comment]
Thanks @KianBadie!  That sounds fine.  Sorry for the radio silence about this.  I've been playing with the demos and thinking about the best way to proceed.  Mol* is awesome and powerful and has a prettier visualization style, but I think it's probably too complex for the student audience.  There's just way too much information there to get lost in.  So I think we should go with PV.  If we do, how much customization can we do?  Can we pick which options to show?  Can we rename the options?  Could we tweak the color schemes?  Can we reverse the zoom direction (at least for me, the mouse scroll button gives the opposite of expected behavior from other sites)?  

I'll give some more thought in the next couple days to the minimal set of options we should include.

**Comment by KianBadie** [comment]
@coiko No worries! I assumed you were giving time to weigh out which viewer fit our needs better. PV it is! To answer your questions: We can pick which options to show. The menu the was included in the PV viewer is actually not built in. The upside to that is that we can style it however we want. The menu I included in the demo was just a quick copy from the [demo](https://biasmv.github.io/pv/demo.html) on the pv site. On the same note of what I just described, we can also rename the options because the menu is created by us. We can tweak the color schemes, as the coloring options have defaults but also take custom color arguments. I am not sure on reversing the zoom, but I hope that we could. That aspect also stood out to me while I was testing things. I will look into that!

Were there other customizations that you had in mind? If it helps, here is a [list of demos](https://pv.readthedocs.io/en/v1.8.1/samples.html) from the pv site that show some of the customization. These demos also include the code to do them, so replicating them should be feasible. One cool demo that I imagine might be useful is [Highlight atom under mouse cursor](https://pv.readthedocs.io/en/v1.8.1/sample-hover.html). I will do some research myself of some other options we have to customize what the viewer can do.

**Comment by coiko** [comment]
@KianBadie Oh that's awesome!  Yes, the hovering over an atom could be very useful.  Thanks!  I'll think about anything else that might be good to have.

**Comment by KianBadie** [comment]
@coiko Sounds good! I will go ahead and start experimenting with PV and see how we can tweak the colors and if it is possible to reverse the scrolling. I'm assuming the details of how PV will be integrated on our site, how it will look, and what options are available will be saved for a more detailed discussion in our Friday meeting?

**Comment by coiko** [comment]
Sounds great.  Thanks @KianBadie!

**Comment by KianBadie** [comment]
@coiko Would you like me to keep the molstar demo on the site to check things out some more? Or are things set and it is safe for it to be removed?

Also, wanted to give a heads up that some I accidentally removed the ability to access the demo viewers when I merged some organization changes today. I will have that fixed soon!

**Comment by coiko** [comment]
@KianBadie No worries and thanks!  And yes, I think it's completely safe to remove the Mol* demo.

**Comment by KianBadie** [comment]
@coiko Sounds great! The demos are now available again and I will go ahead and remove the Mol* demo.

It's a work in progress, but [here](https://www.figma.com/proto/zDsBdxtJXpfXYS7SOV8GbY/PV-Menu?node-id=1%3A2&scaling=min-zoom&page-id=0%3A1) are some simple ideas for what the menu could look like. I'll add more as I think of them, but I wanted to post these ones and see if any of them were in the right direction so that I can free up my hands to start some work on officially integrating the PV viewer into the site. Even though it's a very simple menu, it's admittedly tricky to make a mock of something that is completely satisfying! What do you think of the menus?

**Comment by coiko** [comment]
@KianBadie These are beautiful!  I think the second option, with the transparent background, is perfect.  Thanks so much!

**Comment by KianBadie** [comment]
@coiko I thought that one looked nice too. Glad they worked out!

**Comment by KianBadie** [comment]
@coiko So they way I am going to implement things, I was going to assume anything with a pdb structure id in the video citation is going to have a viewer to open that pdb. What happens in the scenario where there are multiple pdb's in a structure? Or when there is an EMD code?

**Comment by coiko** [comment]
@KianBadie good question!  I think (at least for now) we should ignore EMD codes.  They would require a different viewer and they're less fun for users to play with.  As for multiple PDBs, I can't remember an example.  I can only think of entries that have multiple EMDs, or a PDB and an EMD (e.g. 4.4:Spirosome), in which case we'd just show the viewer for the PDB.  If you encounter one that has multiple PDBs, let me know and I'll figure out a workaround (I can combine them into a single file for display).  Thanks!

**Comment by KianBadie** [comment]
@coiko Sounds good! As for multiple PDBs, I think I was getting it confused with the mixed EMDs and PDBs. We currently don't have an example were there are multiple PDBs, like you said!

**Comment by KianBadie** [comment]
@coiko On our Friday meeting, we talked about how the structure in PV looked different than Mol*. Would you be able to go over the reasons that could be? I tried downloading the PDB file and using that downloaded file as the source, but the difference were still there. One thing that I noticed is that when I switched the Mol* viewer structure type from "assembly" to "model" things started to look similar as the PV demo. For reference, here is [our demo](https://caltechlibrary.github.io/cell-atlas/pv-demo) and the [Mol*](https://www.rcsb.org/3d-view/6UEW) representations of PDB 6UEW.

Also, holding shift and using left click on PV pans the camera! That was a pleasant surprise. Maybe our "detect right-click" tooltip can advise to use shift to pan the camera.

**Comment by coiko** [comment]
@KianBadie Woohoo!  That's awesome (the panning feature)!

And yes, the different appearance in PV and Mol* should be easily fixed.  I think they're using different files from the PDB.  The one we want (and that Mol* probably uses) is called "Biological Assembly 1".  (Basically it has the whole functional complex, rather than just the smaller repetitive subunit, which is what's in the default PDB format file.)  I'm attaching that file here - see if it does the trick.  In many cases, the two files are the same, but there are a few special cases like this one, just to keep us on our toes.
[6uew.pdb1.zip](https://github.com/caltechlibrary/cell-atlas/files/6619920/6uew.pdb1.zip)

**Comment by coiko** [comment]
@KianBadie Wonderful!  Yes - they will all be that "Biological Assembly 1" download.  There are 3 file formats (.pdb, .pdb1, and .cif) that are all opened by the same programs (.pdb1 just indicates that it's the assembly (i.e. there's more than 1 subunit) and .cif is used for larger files) so they should all work with PV, but let me know if they don't and I can convert.  I sent you a zip with all the files, if that's easier.  (Download link: [here](https://dropbox.caltech.edu/cgi-bin/dropbox.cgi?get=1&key=be1b09c243646233ff920a65e55ab32a)). Let me know if you find anything's missing or run into any trouble with anything.  Thanks!

**Comment by KianBadie** [comment]
@coiko Sounds good! I will let you know if I run into any issues. So far I have not, and Tom said they should all work as well so I think we should be ok! And thank you for the zip, that will make things easier. I will let you know if anything else is needed!

Also, the development on the viewer has been thankfully smooth so far! I attached a gif of how things are looking so far. It is an exciting feature to work on!

## Gif
![viewer-demo](https://user-images.githubusercontent.com/18221058/121404324-b64b9b00-c910-11eb-8c31-88090d53cabc.gif)

**Comment by coiko** [comment]
Oh how cool!  Thanks @KianBadie!

**Comment by KianBadie** [comment]
@coiko It looks like PV is having some difficulty working with the cif files (the pdb/pdb1 files are working well though). I've been trying to fiddle with PV to see if there is something I am missing but it is not looking to promising. There is built in functionality for sdf files, but I'm not seeing anything for cif files. Unfortunately, it looks like there is not too much discussion on this in the PV documentation or github issues. you mentioned converting in your earlier comment. Is there a convenient way to convert while still displaying the same thing? Or does converting change the intended model to show? I'll check in with Tom as well. I'm not sure how extensively he knows PV, but I will see if this sounds familiar to him to double check if converting is necessary.

For my curiosity, are cif files formatted differently from a pdb/pdb1 file? Or is it just a bigger pdb file? Because if it is the latter, I'm not sure why yet it doesn't work.

**Comment by coiko** [comment]
@KianBadie Ah, sorry to hear that, but no worries.  I'll try converting it tomorrow; I think it should be easy because as far as I know the base structure is the same.

**Comment by coiko** [comment]
@KianBadie Update:  the file conversion seemed to work fine from what I can see, so I just sent you a new .zip with all the files as .pdb1 (download link also [here](https://dropbox.caltech.edu/cgi-bin/dropbox.cgi?get=1&key=cad1e6a85f63027775b4fc4bea8cce2d)).  Let me know if anything doesn't work.  Thanks!

**Comment by KianBadie** [comment]
@coiko When I get around to merging the protein viewers today, are you ok with the pv demo being removed along with it? Or would you like to keep it for some time?

**Comment by coiko** [comment]
@KianBadie Yes, I think it's great to remove the demo - thanks!

Also, here's the missing PDB file for 3JC8.  Let me know if you find (or don't find) anything else!
[3jc8.pdb1.zip](https://github.com/caltechlibrary/cell-atlas/files/6640609/3jc8.pdb1.zip)

**Comment by KianBadie** [comment]
@coiko Sounds good! And thank you for the file!

There was actually an unfortunate problem with the cif converted pdb files. It seems that PV renders everything fine on my computer using "secondary structure" model. However, when I try to change things to the "volume filling" model, it looks like my browser takes a performance hit and eventually gives up and giving an error message related to rendering context (this would happen even after configuring PV to low quality settings and with no antialiasing). From some research, I'm guessing the pdb in question (6o9j) is too complex for my computer to render quickly so the browser is killing the long process. When I try this with another cif converted pdb (5tcr) things worked ok on my desktop, but not on an iPhone 8. Safari on iPhone 8 would eventually display a message replacing the whole page saying "A problem repeatedly occurred on <url>", which I feel like makes me believe the performance theory more.

I got in touch with Tom, and he agreed that it sounds like a performance issue. He said it is probably too difficult to re-engineer PV to be more efficient, so he gave three possibilities. In his words in his reply email to me:
1. Remove the sphere model as an option for the large proteins
2. Make a modified pdb file that removes the atoms of the protein that you use for just the sphere model. Since the sphere model only shows the outer surface of the protein, you can often get very similar results using fewer atoms. For instance, you could include only the CA atoms in the pdb file. This will give you an approximate surface, but won’t work for the other visualization types. You could also strip out atoms that are at the center of each subunit, but there isn’t a great automated way to do this.
3. Use Mol* for the large proteins.

For option 2, I just told Tom that "ball and sticks" model was also giving problems, so I will see what he says about that option considering that (since he said that option won't work for other visualization types). I tried playing around with a couple of similar looking model types for pdb 6o9j in Mol*, and things seemed to render ok. Another thing that I wanted to look into that might be important is that the error message for the rendering context is detectable, so we could catch that error and do something in response. However, I am not sure if it alleviates the experience of the performance hit. On my computer, the current tab with the viewer will freeze for a bit, then the whole browser will go black for a second or 2, and then things will appear again with a blank canvas and the error message in the console.

My apologies for this roadblock coming right before we thought it was a wrap for the protein viewer 1.0. With all this in mind, would you still want a merge for today? Or wait until he have a solution for the large PDBs? My guess is the large PDBs might still be viewable on the default "secondary structure" model (depending on the computer). However, I would understand if it is better to hold off because of this issue.

**Comment by coiko** [comment]
@KianBadie Thanks for this great explanation, and I'm sorry you ran into that problem.  I don't think it's worth waiting to merge while we figure this out, so go ahead if you'd like (but if you'd rather wait, that's totally fine, too!).

That particular structure is monstrously big.  I'll give some more thought to the best way to handle large PDBs.  We could limit the visualization style, or we could show just a piece of the whole structure, or we could just not enable the viewer for problematic cases.  Anyway, I'll ruminate on it over the weekend and let you know any thoughts that (I hope) occur to me.  Thanks!

**Comment by KianBadie** [comment]
@coiko No problem and no worries! I'm afraid I might have to take up your offer to wait, because there was actually another unfortunate problem. It looks like GitHub is having some complaints about the file sizes of the pdb files when I try to push them to our repository. I had to delete pdb 3j31 because it was over the file size limit, and I think 6kgx was the one causing the complaint after that. Things could work out if we put things on our main server and call the files from there. Tom said he could do that Monday, and I was thinking a possibility is to use the urls on the pdb website (since PV can fetch files remotely). However, even though it is a minor task to implement that change, I think I should just call it a day and save it for next week since I would prefer to not push changes to our testing site without sitting with them with a clear mind for a bit if the option to wait is there! In addition, I would still have to hide the modal when the viewer is open, which is minor but a similar situation to what I stated above. Again, my apologies for not having the viewer merged today after promising it in our meeting today. I will plan on having the protein viewer merged next week after letting things sit for a bit, hiding the modal, and changing things to use the pdb website urls or our server urls.

Ah I see! Yeah, I think those are all options that are considerable. One other option I thought of today was maybe even putting a warning on problematic rendering options for the large files. Maybe a pop up message could show up on hover on the render options with a "render at your own risk of performance hit" kind of message? And no problem! 

Lastly, as a follow-up to solution 2. in my previous comment (where I was not sure how that would apply to "ball and sticks" rendering), Tom told me it would work (functionally) with balls and sticks, but you wouldn’t be able to see all the atoms since they are not in the file. If you want the balls and sticks representation on the large models, he recommends to probably do two views with two pdb files (one with the whole complex and one with just the area or interest).

Again, sorry for all the last minute changes in plans. Hopefully things can be merged in smoothly next week. Until then, hope you have a good weekend!

**Comment by coiko** [comment]
@KianBadie No worries at all!  That sounds great, and it's definitely a good stopping point for the week.  We'll figure out how to handle the files next week.  In the meantime, have a great weekend!

**Comment by KianBadie** [comment]
@coiko The protein viewer is now up on our testing site! Again, my apologies for the delays and new problems we found along the way. One interesting thing to note is that the pdb files behave similarly to the video files in that access issues might arise when switching between urls (testing site/main site). This might not be a problem for you since the only place for you to access the viewer outside of building the site and serving locally is the testing site. But I just wanted to get that out there now.

On the topic of performance, I haven't reached any "aha" moments but I have had a new thought. I think one thing that adds to trickiness of this issue is that I imagine it is also dependent on the device being used. I would have to do some more testing, but what if mobile devices have trouble rendering a certain pdb while desktop doesn't have a problem. If that is true for some cases, then it makes the issue a little more unpredictable.

Also, here is a gif of the iOS behavior I was talking about. I thought I would include since you have an android. There is a weird jump that happens when the color/model selector opens up. It might look laggy in the gif, but things are smooth on my actual iPhone (and look cool as well!). Other than that, I think I covered the important bits about iOS in our meetings/this gif:
![viewer-ios](https://user-images.githubusercontent.com/18221058/121944378-779a5400-cd07-11eb-86b7-450c914f5db4.gif)

**Comment by coiko** [comment]
Thanks @KianBadie Sorry for the slow response, but this is fantastic!  It looks really great!  I agree with your concern that the performance issues will likely vary by platform.  I had an idea about how to shrink the large file sizes in a way that doesn't compromise the biological information too much, but it might take me another day or two to see if it will work.  I'll keep you posted on whether that looks promising.

**Comment by KianBadie** [comment]
@coiko No worries on the response time, I know things like this take time to sit on. And sounds good! I will let you know if I run into any new ideas on this issue as well.

**Comment by coiko** [comment]
@KianBadie Thanks for your patience on the large-file issue!  My idea didn't work out, and after more thought, I think the best solution is just to limit the view options based on size.  For larger files, we can offer only "secondary structure" and "backbone" (I think those should be about the same complexity for rendering, but I might be wrong).  Also, we may want to further limit options on mobile - maybe we should only show secondary structure for all of them?

Assuming that works, the other issue is the size of the files themselves.  Will that be a problem on the main site, or just in the repository for testing?  Pulling the files from the PDB runs into the file format problem, since most of those "Biological Assembly" files are .cif format.  If we can get a good idea of what size becomes problematic, I can make .pdb files for the larger structures that only contain the coordinates of atoms in the backbone.  In a quick test, that seems to cut the file size by about half.  (But with that file, we can only display the backbone view.)  What do you think?

**Comment by KianBadie** [comment]
@coiko No problem! And for the time being, I think that is a fine solution to this tricky problem. Maybe we can start with "secondary structure" and "backbone" and see if we still run into any issues. I think it wouldn't hurt to try both on mobile as well and see how things go. What are your thoughts on that? I guess the next thing would be to determine what is qualified as a large pdb. Maybe we could do some testing to find the which pdb breaks the threshold. From a quick look of things now, file size for the pdb files climb gradually from under 1MB to around 10MB until the last three big pdb files, which are 5u3c, 6kgx, and 3j31 (topping out at 200MB). So I can do some investigating to see where the cut off point lands.

My apologies, I never gave a direct update on the file size issue. On Monday, Tom uploaded all of the pdb files that you sent me  to the same server that hosts the main site. And our testing site is getting the pdb files from that server in the same way that we get the videos. So the large file size is no longer a problem for the testing site and will also be accessible on the main site. One catch is that the same cross origin issues come up like the videos if one is flipping between sites.

**Comment by coiko** [comment]
That sounds great @KianBadie!  

Thanks to you and Tom for setting up the files for the testing site.  The structures weren't showing up for me this morning for some reason but they look great now (until I make it stick by trying to play with space-filling models of the biggest ones ;)  Thanks!

**Comment by KianBadie** [comment]
@coiko Awesome! I will let you know any important findings as they come.

No problem! And please let me know if that problem happens again and I will investigate. It is possible that there could be a bug that prevents the files from being accessed in certain situations!

**Comment by coiko** [comment]
@KianBadie I played around a bit with a potential icon.  What do you think about either of these (or something else)?

![PVicon_box](https://user-images.githubusercontent.com/56415464/122495002-8babb300-cf9e-11eb-8945-6a4d351f36e6.png)
![PVicon_nobox](https://user-images.githubusercontent.com/56415464/122495046-8e0e0d00-cf9e-11eb-878c-3a0a44ececd0.png)

**Comment by coiko** [comment]
@KianBadie And one more...with the box but horizontal as we discussed.  Let me know how it works.  Thanks!

![PVicon_box_horizontal](https://user-images.githubusercontent.com/56415464/122601369-0bc92b80-d026-11eb-892b-5d22cd678875.png)

**Comment by KianBadie** [comment]
I tested out the viewer icons. I think the horizontal ones are clearer than the vertical one. Seeing them in the context of the modal, how do you think they look? I'm wondering if the video citation element is too small for the image to be clear on what it is. Do you think the buttons could be bigger in the video citation element?

## Screenshots
![vertical](https://user-images.githubusercontent.com/18221058/122624389-cec85d80-d054-11eb-8af5-80256505cd1e.PNG)
![horizontal](https://user-images.githubusercontent.com/18221058/122624388-ce2fc700-d054-11eb-8e35-f65211d02201.PNG)
![3d-horizontal](https://user-images.githubusercontent.com/18221058/122624387-cd973080-d054-11eb-8aed-b95ec5f4c14c.PNG)

**Comment by coiko** [comment]
I think it looks great!  Thanks @KianBadie!  I agree that horizontal is better, and I like the boxed version (your last screenshot) best of all.  I might just move it slightly farther to the right, maybe three times as far from the PDB link as it is now, so it's clearer that it's separate.  We'll highlight it in an updated intro video for the site (along with the other awesome features you've added), so I think it will be clear what it does.  Maybe at some point for super-accessibility we could also add a hover tip that would say something like "interactive viewer".  Of course, the user can also just click it and find out!

**Comment by KianBadie** [comment]
@coiko For sure! I will go ahead and merge that small update into the site and we can further see how things look in action! I think the hover tip is also a great idea!

**Comment by KianBadie** [comment]
@coiko I just merged the icon into the site! To make the display of the video citation elements look nice with the protein viewer button, I had to make some small changes. A side affect of that is that the "Diversity" subsection modal has different line breaking in the video citation (things are more centered). From my memory, this is the only edge case with the all the others being the typical doi/structure dois, and I thought that those looked good. I figured everything still looked decent enough to throw it in today. If you think anything should be tweaked more or changed back please let me know!

## New
![new](https://user-images.githubusercontent.com/18221058/122626937-ab57df80-d061-11eb-8f8d-6dfe002c7b11.PNG)
## Old
![old](https://user-images.githubusercontent.com/18221058/122626941-aeeb6680-d061-11eb-8349-071c9a78017b.PNG)

**Comment by coiko** [comment]
@KianBadie The icon looks fantastic!  And this crazy list looks even better than before.  Thanks!

**Comment by KianBadie** [comment]
@coiko I have ran into some interesting findings on PV. Up until now, I have been using the latest release of PV (v1.8.1). However, I learned that there is a v1.9 that is not labeled as a release, but is available to download. One of the additions on this version was changing the way the spheres model is rendered. Switching to this version did indeed allow sphere models to render more without crashing on desktop. Even on the largest file size pdb 3j31, my computer was able to render it with spheres (my computer, while a little old, still seems like it has more power than an average computer). Unfortunately, the Mac I tested things on was not able to render 3j31 at all (any model). That is our biggest pdb however (almost 2x bigger than the 2nd biggest if I remember correctly), so hopefully that is not the norm for other proteins on desktop. Besides spheres modeling on dekstop, everything else rendered about the same.

If we do want to use this version, here are some other noticeable difference/addition:
- Inverted color of rainbow gradient.
- Support pan by middle-click.

There was another interesting finding that I found. While I knew about the viewer configuration options to help with performance (quality and antialiasing), I overlooked that there are also config options to pass on a rendering type level. Spheres has `sphereDetail` option. Balls and sticks has `arcDetail` and a `sphereDetail` arguments. I am going to test how tweaking these works out and if they can help with increasing the range of devices that can support different modeling types.

**Comment by KianBadie** [comment]
@coiko I incorporated an initial version of the atom highlighting and have attached a gif of it below. What are your thoughts on it so far? Do you think any particular styling could enhance it? I defaulted to a red highlighting like PV's demo. However, on the rainbow coloring, the highlighting doesn't show on other red structures. Should the highlight be an outline maybe? I can see if PV can do that.

## Gif
![viewer](https://user-images.githubusercontent.com/18221058/122847333-549a1200-d2bc-11eb-85ad-a90b3fbcc51c.gif)

**Comment by coiko** [comment]
@KianBadie Thanks so much for looking into all this!  The new version sounds like an improvement.  I'm not sure it's worth keeping the ball-and-stick option for large files, though.  It's cool that it works, but it looks pretty confusing to me at the lower resolution.

The residue highlighter looks fantastic!  I love the styling!  And I think your idea of of an outline instead of a color would be perfect!

**Comment by KianBadie** [comment]
@coiko No problem! I will go ahead and update the version then. Yeah, the lower resolution definitely does not look as good. I can tweak it some more and see if it yields good results.

And sounds good! I will go ahead and implement that then!

**Comment by KianBadie** [comment]
@coiko Unfortunately, I was not able to find a way to outline certain selections of the protein in PV. Do you have any other ideas? Would the coloring the selection black be a suitable choice?

**Comment by coiko** [comment]
@KianBadie No worries!  That does sound pretty advanced now that I think about it.  Black would be great, or white (since it still shows the black stroke on the sides) - I think either one would stand out nicely.  Thanks!

**Comment by KianBadie** [comment]
@coiko Sounds good and no problem!

**Comment by KianBadie** [comment]
@coiko I pushed the atom label updates today. They are available to use on mobile and desktop. On desktop, I used to have the label/highlight display when you hovered over a structure. However, I think this created a performance hit as PV needed to redraw every time you wanted to highlight a part of the structure. This was fine on smaller proteins, but the lag was noticeable when you would move your mouse around. The highlight wouldn't register until the mouse was still. So I changed it to be a click to highlight functionality. If you're interested in the hovering functionality, let me know and we can test it out and see if the performance hit is acceptable (since it was not as bad as the trying to render balls and sticks on huge proteins).

**Comment by coiko** [comment]
Thanks @KianBadie!  It looks great - I like the black highlight.  I also like the idea of clicking rather than hovering to highlight a residue.  I think it's good to keep the computational demands as light as possible.

**Comment by KianBadie** [comment]
@coiko Like I said in our meeting today, I have the functionality to limit rendering options ready to merge. However, I wanted to check to see if you wanted to take any more time to check things out before I limit any rendering options. I just wanted to make sure I wasn't limiting options before giving a chance to your devices to see if they could handle them. I'm still fielding things out, but right now I still have the threshold at 5MB and taking away just ball/sticks on desktop and ball/sticks + spheres on mobile. 

If you were ok with merging whatever starting point we have now, I can merge things and we can try increasing/decreasing the file size/options from there. What do you think?

**Comment by coiko** [comment]
@KianBadie Thanks so much for checking, but I've gotten a chance to play around with the current options, so go ahead and merge in your awesome limiting function!  And I think those parameters sound perfect.

**Comment by KianBadie** [comment]
@coiko No problem and sounds good! I will go ahead and merge them. And I'll actually continue the protein viewer stress test next week, as I think I am protein viewer-ed out for the day! Next week I will see if things should be increased or decreased and I'll accept the 5MBs and removing both the ball/sticks + spheres option (but keeping spheres on desktop) for now!

**Comment by KianBadie** [comment]
@coiko It's merged! Please let me know if you experience any issues

**Comment by KianBadie** [comment]
@coiko I added HOH to the look up dictionary, but please let me know if you encounter anything else that gives a wonky label in the protein viewer!

**Comment by coiko** [comment]
@KianBadie Perfect - will do!  Thank you so much for getting this merged in!  I'll have fun playing with it and let you know if I find any issues.  In the meantime, have a great protein-viewer-less weekend!

**Comment by KianBadie** [comment]
@coiko I attached a gif below of what the new gradient looks like on a random pdb. It only applies to the "Individual Proteins" option, and the colors on the other 2 options are the same as before. I am not sure if there is a specific way the gradient is supposed to be carried throughout the protein. Does the below gif look as expected? Or is there something different from expected?

As a side note, the new color gradient is very nice and easy on the eyes.

## Gif
![gradient](https://user-images.githubusercontent.com/18221058/128074335-be0032e6-b213-4de1-a0c4-ed49e264f606.gif)

**Comment by coiko** [comment]
@KianBadie Oh beautiful!  It looks perfect!

**Comment by KianBadie** [comment]
@coiko Awesome! That is great to hear. Similarly to my above question, how does the secondary structure look?

## Screenshot
![ssGradient](https://user-images.githubusercontent.com/18221058/128076540-3b8107de-d4e1-4ee8-af87-f67d98aeb87b.PNG)

**Comment by coiko** [comment]
@KianBadie It looks great!  It's coloring exactly what it's supposed to color.  I think it might be better, though, to reverse the blue and orange because most proteins have many more helices (currently orange) than strands (currently blue) so the orange would make the rarer element stand out.  If it's easy and you don't mind trying it, could you flip the two colors?

**Comment by KianBadie** [comment]
@coiko Not a problem, I flipped the colors and pushed the changes to the testing site. they should be reflected their soon. Please let me know if anything needs changing or if something was implemented not as intended!

**Comment by coiko** [comment]
@KianBadie Thank you so much!  It looks perfect!

**Comment by KianBadie** [comment]
@coiko I merged the protein viewer updates to the testing site. Just in case you don't remember all of them, the updates were:
- Dropped depth coloring
- Change the cursor for the protein button
- Display "water" as just "water" in the protein label

Please let me know if you notice anything else!

**Comment by coiko** [comment]
Beautiful - thanks @KianBadie!  Those changes look great!  Unfortunately, I'm getting the non-loading problem again with a few of the structures (most noticeably the ones in Chapter 2).  Here's the error (from Chrome):
<img width="527" alt="Screen Shot 2021-08-09 at 1 12 39 PM" src="https://user-images.githubusercontent.com/56415464/128768471-5e73e28e-995a-49c3-bc20-71b43b26384d.png">

Also, on an unrelated (except related to the protein viewer) note, I noticed that Safari doesn't handle the zoom function as well as Chrome.  At least for me, it zooms the whole page rather than just the structure.  Is there a different way to zoom it in Safari, and can we copy the Chrome style?  (Also, I have to say again that the behavior on mobile is awesome!)

**Comment by KianBadie** [comment]
@coiko That is odd that you are getting that error. This might be a reach, but I just wanted to make sure: have you recently built and served the site from your own computer? If not, then there might be a trickier issue going on behind the scenes that I will get in contact with Tom about.

Oh that is also interesting, I will try to replicate that and investigate what is going on and try to make things consistent. That is awesome to hear that mobile is doing well!

**Comment by coiko** [comment]
Thanks @KianBadie!  I haven't built/served the site for quite a while (certainly not since you added the protein viewer), so it might be something behind the scenes?  Thanks for looking into it before we move things onto the main site!

**Comment by KianBadie** [comment]
@coiko I see, I will do some digging and try to find out what is going on. On the main site, I don't think this will be an issue since cross origin errors don't happen since it is on the same server. But I still want to get to the bottom of this since it is a real problem. I will let you know of any findings.

Also, I tried to replicate the scrolling/zooming behavior and was not able to do so. Was this on mobile or on desktop? And when zooming, is it by just using the mouse wheel or are any other buttons being used?

**Comment by coiko** [comment]
@KianBadie Ah, sorry, I left out some critical information!  It's a Mac laptop (or trackpad, I guess)-specific problem.  A mouse with a scroll wheel works great.  But on the trackpad, if you do the pinch-expand gesture, on Chrome it has the same behavior as the mouse wheel (zooming just the structure), but on Safari it zooms the whole page.  Don't worry about it if it's not an easy fix - we can certainly leave it to the future, but I just wanted to bring it up while it occurred to me.  Thanks!

(Also feel free to ignore the cross origin error on the protein structures since it won't happen on the main site!)

**Comment by KianBadie** [comment]
@coiko No worries! Ah I see, that's interesting that the two treat the same device differently. And sounds good! In that case I will lower it on my priority list for the next update and continue to work on other things, but will come back to it in the future. 

Sounds good as well with the cross origin issue!

**Comment by KianBadie** [comment]
@coiko I think I figured out why the PDB numbering was weird on some proteins. As you know, the protein number is determined by the qualified name (for example "J.TYR83.CA") and that the first character is what determines the number. However, I found that a protein with a weird protein number had the qualified name "3.VAL93.CA". The way I wrote things, the label assumed only alphabetical characters would be there but did not account for numbers.

Are numbers an expected thing in the qualified name for the protein? Or is this a PV thing? If it is a PV thing, I can try to investigate how they arrived at giving a number for the protein. If it is expected, where do numbers fall on the protein number ordering? Is "1" before "A" and what would numbers translate to?

**Comment by coiko** [comment]
@KianBadie Ah, that makes sense - thanks!  Yes, unfortunately numbers are an expected identifier for proteins (along with letters).  When structures get really big (e.g. like 6kgx), they use capital letters (A-Z), then lowercase letters (a-z), then numbers (1-9) to identify up to 61 proteins. (If even more identifiers are needed, they use a two-character system, doubling them up in various configurations, like A1-9, B1-9...a1-9...Aa-z... but we shouldn't have to deal with that.)  Sorry I forgot to tell you about this earlier (I assumed PV was handling the numbering)!

When you have the time and inclination to fix it (feel free to wait - this is very low priority!), the translation should be: A-Z = 1-26, a-z = 27-52, 1-9 = 53-61.  (As a further wrinkle, some structures skip some intervening letters/numbers so you'll have e.g. A, B, C and F, so if you ever feel really ambitious, you could also clean those up to consecutive numbers, but that's really not necessary.)

**Comment by KianBadie** [comment]
@coiko Ah I see, and no worries! If it is that low priority, I will let you know if I get around to it. I will probably just create a look up table for that. I think accounting for the skipping would for sure have to wait for the next update, as that will take a little bit more time/logic.

Also, I pushed some changes that improved the protein highlighting. At least on my computer, the selected protein seems to follow the tip of the pointer. Unfortunately, it did not look as accurate when I tested in Safari, but it did look like it improved from the previous version. Please let me know if things seem the same to you. The problem ended up being on my end. I'm using the built in PV functionality that selects a residue when given a certain position. The clicked position I was giving was not accounting for the borders of the viewer.

**Comment by coiko** [comment]
@KianBadie Thank you so much!  It looks perfect to me now in Chrome, and much, much better in Safari.  

(And yes, I think it's great to wait on the numbering issue.)

**Comment by KianBadie** [comment]
@coiko I uploaded some changes to use a look up table that covers A-Z, a-z, and 1-9. It looks like it is being used as intended, but something still look off. For example, in the 4.3 PDB, some of the proteins are numbered as 56. I'm not sure if it is correct, if there is something with the file, or PV because 56 seems like a high jump when the other proteins sit around the 10s area. If you notice any thing that looks off, please let me know!

**Comment by KianBadie** [comment]
@coiko In addition, I noticed that PDB 609J had "undefined" for the protein name because they did not exist in our protein look up table. Some of the names that were not in the table were "N1", "C3", and "N3". One small/easy thing I could do is to just not display a protein name for proteins that aren't in our look up table. Another option is to add more entries into the table. What are your thoughts?

**Comment by KianBadie** [comment]
@coiko I was testing the protein viewer on iOS and had some new thoughts on the full-screen functionality that it has. For a refresher, iOS does not have true full-screen for non-video elements, so we made our own implementation. Because of this however, the nav bars get in the way of the top/bottom of the full-screen elements sometimes. In the past, I added some functionality to resize the element when the nav bars come in/out for the comparison slider, but as we discussed that causes bugs with zooming so we removed it for future elements. The protein viewer is included in this.

However, I did notice that on portrait, it was a little bit jarring for the nav bars to pop up when I was trying to click the "X" in the bottom left corner, and then covering that element (so I had to switch to landscape to actually exit the viewer). One potential "fix" that I thought of was to move the "X" to the top right corner on mobile only. Most of the full-screen elements on mobile have the "X" in the top right, while having it in the bottom right for desktop (with an exception for the image elements in the modals). So making this change would somewhat add consistency. What are your thoughts?

Ideally in the future, I would make things in a way where this bug wouldn't happen in the first place. However, I think now is a little too late to introduce functionality like that without fully testing the side affects.

**Comment by coiko** [comment]
@KianBadie sorry for the delayed response, but thanks so much for all this!

As for the numbering, that PDB in 4.3 (6kgx) is a beast and uses the combo two-digit/letter system (basically every combination of upper/lower/numbers) to label a few hundred different proteins.  So when I said "but we shouldn't have to deal with that" I was very wrong and had totally forgotten about how big this thing is.  I'm sorry!  That said, I definitely wouldn't try to fix it now.  I think we can look into it down the road and for now, the main thing is that the residue selector works.

As for the undefined residues in 6O9J, good catch!  I had also forgotten about them.  It's a weird hybrid structure that's part protein and part RNA (a different macromolecule).  The residue selector only makes sense for the protein part of it, so yes, I think the fix you proposed of just not identifying anything not in the lookup table is the perfect way to handle it.  Thanks!

Finally, for the full-screen functionality, I really like the idea of moving the "X" to the top right corner on mobile.  (As an added bonus, I think the top-right corner is more intuitive from other sites.)  So yes, please go ahead with that.

**Comment by KianBadie** [comment]
@coiko No worries and no problem!

No worries and it sounds good to me to save the numbering fix for later.

Sounds good. When you say "not identifying anything not in the lookup table", does that mean no label at all? At first I was thinking of displaying "Protein 2 | Residue #5 | undefined" as "Protein 2 | Residue #5", but your explanation makes it seem like there shouldn't be anything displayed.

I will go ahead and move it to the top right!

**Comment by coiko** [comment]
Thanks @KianBadie!  And yes, I think it would be best not to display anything since it's not actually a protein (unless it's more of a pain to set it up that way, in which case we can definitely go with "undefined").

**Comment by KianBadie** [comment]
@coiko Now that I think about it, do you think it would be fitting if something was displayed? I ask because I was about to implement showing nothing at all, but then I thought of the scenario that somebody clicks something, the structure highlights, but no label shows. I was thinking it might come off as a bug. Then I thought that if those structures were not highlightable, then it might also give off a wrong impression. What are your thoughts on displaying some sort of phrase that denotes that the highlighted structure is not a protein? Or do you think it would not fit with the purposes of the viewer?

**Comment by coiko** [comment]
@KianBadie Yes, that's a great idea!  Maybe it could be the same style as the "water" label and just say "non-protein".  What do you think?

**Comment by KianBadie** [comment]
@coiko I think that sounds great! "non-protein" it is.

**Comment by KianBadie** [comment]
@coiko I just pushed some changes to display that custom "non-protein" message if the highlighted atom does not exists in our protein look up table. Please let me know if you notice any future issues with that!

**Comment by coiko** [comment]
@KianBadie Wonderful!  Thank you so much!

Also, I just talked with Grant and it looks like the viewer doesn't show up in Firefox (it just shows a funny semi-transparent window with the display options at the top).  So maybe we can add troubleshooting that to the to-do-sometime-but-not-necessarily-before-this-update list?

**Comment by KianBadie** [comment]
@coiko No problem! I also just pushed some changes to remove ball and sticks/volume filling rendering on mobile.

And oh I see that is unfortunate if PV and firefox do not work with each other. Yes, that sounds good to me. I'll do a quick check right now and see if something else is going on.

**Comment by coiko** [comment]
Thanks @KianBadie!  It looks great!

**Comment by KianBadie** [comment]
@coiko I ended giving some more time to the bug Grant was experiencing and just pushed some changes now! It looks like the problem was something silly. I was using a property that only Chrome and Mac understood to calculate the size of the viewer. Because Firefox did not understand it, it rendered the viewer with no dimensions.

**Comment by coiko** [comment]
@KianBadie Oh wonderful!  Thank you so much!
